### PR TITLE
Support Azure AD authentication: Implementation

### PR DESCRIPTION
**Pull request by jeliasson**
This PR seeks to implement Azure AD as authentication in the `@redwoodjs/auth` package.

See https://github.com/redwoodjs/redwood/issues/1310

### CODE REVIEW

File: packages/api/src/auth/decoders/azureAd.ts
Comment (dthyresson) [inline packages/api/src/auth/decoders/azureAd.ts:4]: From reading some [Azure AD docs](https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens#validating-the-signature), the verification could be similar the Auth0's which is implemented by getting the signing key via a jwksClient:

https://github.com/redwoodjs/redwood/blob/main/packages/api/src/auth/decoders/auth0.ts#L25

File: packages/api/src/auth/decoders/index.ts
Comment (dthyresson) [inline packages/api/src/auth/decoders/index.ts:12]: Not certain the Java/Typescript naming convention:

* azureAd
*azureAD
* azureActiveDirectory
?

File: packages/cli/src/commands/generate/auth/templates/azureAd.auth.js.template
Comment (dthyresson) [inline packages/cli/src/commands/generate/auth/templates/azureAd.auth.js.template:24]: It **looks** like AzureAD [supports roles](https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens#user-and-application-tokens) with a `roles` claim on the token.

If so your template may want to include the "role" version of require auth:

https://github.com/redwoodjs/redwood/blob/main/packages/cli/src/commands/generate/auth/templates/auth.js.template#L104

`parseJWT` should still find those with the `roles` claim at same level as `sub`:

https://github.com/redwoodjs/redwood/blob/main/packages/api/src/parseJWT.ts#L11

```
  return (
    token.decoded?.roles ||
    metadata?.roles ||
    metadata.authorization?.roles ||
    []
  )
```

File: packages/api/src/auth/decoders/azureAd.ts
Comment (jeliasson) [reply packages/api/src/auth/decoders/azureAd.ts:4]: Thanks @dthyresson! I'll have a closer look later this week :)
Also, I have marked the PR's as a drafts to make life easier for maintainers.

File: packages/api/src/auth/decoders/index.ts
Comment (jeliasson) [reply packages/api/src/auth/decoders/index.ts:12]: @dthyresson Not me either. Personally I would go with `azureAD` but I'm not sure that is the right convention. `azureActiveDirectory` is a bit to long imo. Can we get input from @peterp in this maybe?

Once decided, I'll make that change to all references.

File: packages/cli/src/commands/generate/auth/templates/azureAd.auth.js.template
Comment (jeliasson) [reply packages/cli/src/commands/generate/auth/templates/azureAd.auth.js.template:24]: @dthyresson Agreed! I will give this a go along with JWT validation. Thanks for the review and references!

File: packages/api/src/auth/decoders/index.ts
Comment (peterp) [reply packages/api/src/auth/decoders/index.ts:12]: Thanks @jeliasson :)

I would personally go with the really long name. Since people who are unfamiliar with it then have a full reference instead of "AD"

**Comment by peterp** [comment]
@jeliasson You've done amazing work here, thanks @dthyresson for reviewing this :)

File: packages/api/src/auth/decoders/azureAd.ts
Comment (jeliasson) [reply packages/api/src/auth/decoders/azureAd.ts]: @dthyresson I will mark this as resolved (as the new file is called `azureActiveDirectory`) which contains validation. I'll ping you in a new reply.

File: packages/cli/src/commands/generate/auth/templates/azureAd.auth.js.template
Comment (jeliasson) [reply packages/cli/src/commands/generate/auth/templates/azureAd.auth.js.template]: Marking as resolved due to outdated.

/cc @dthyresson

**Comment by jeliasson** [comment]
Hey all 🖐

@dthyresson I have implemented [token validation](https://github.com/redwoodjs/redwood/pull/1311/commits/9b80e7e5d34f9ecae9c002947488a94e6e3299cd) (using Auth0 decoder example, thanks for pointing me in that direction!) and a better [auth template](https://github.com/redwoodjs/redwood/pull/1311/commits/ba10199713a709c9930ba7bcc04ec2f28d9ee83c#diff-0e80ed6b9b02457ed52349a91c76773df1c26f1b3df76d435eaffaa2b2e8638c) for Azure Active Directory which picks up the [roles nicely](https://github.com/redwoodjs/playground-auth/pull/6#issuecomment-723657903). By any chance you can have another review at your convenience? 🙏

@thedavidprice By any chance you can estimate time on the `setup` command and migration from `generate` as of #1307 #1309? Also, is there any chance to maybe merge this into a canary release? I'm about to start a new project based of Redwood, and that project will require this authentication method. 😊 

Finally, is there any "strategy" what various token claims that should be surfaced to `currentUser`? In [the auth generator](https://github.com/redwoodjs/playground-auth/pull/6#issuecomment-723657903) I bring in everything from `...decoded` and adding `email` from `decoded.prefered_username`. Thoughts on this?

/cc @peterp

**Comment by Tobbe** [comment]
@jeliasson I'm sure @thedavidprice wouldn't mind some help with #1307 and #1309. If you want to jump in and do some work on those I'd be happy to review/merge to give you a canary to try 🙂

**Comment by jeliasson** [comment]
> I'm sure @thedavidprice wouldn't mind some help with #1307 and #1309. If you want to jump in and do some work on those I'd be happy to review/merge to give you a canary to try 🙂

Unfortunately I'm swamped with work this month, otherwise I would catch up on those and jump right in. I would like to focus on the authentication, test in a real-life application from a canary, and I'll be happy to contribute to `setup` in December. 📌

If canary is not possible, is there any nice way of patching something under `node_modules`? I.e. manually bring in the changes. Thanks!

**Comment by Tobbe** [comment]
> If canary is not possible, is there any nice way of patching something under node_modules? I.e. manually bring in the changes. This would help greatly in the project I'm about to spin up. Thanks!

@jeliasson You could voice your support the the issue I just created :) #1458

**Comment by andrew-hwahin** [comment]
Hope this will be released soon😁🙏🙇‍♂️

File: packages/api/src/auth/decoders/azureActiveDirectory.ts
Comment (Tobbe) [inline packages/api/src/auth/decoders/azureActiveDirectory.ts:11]:
```suggestion
        '`AZURE_ACTIVE_DIRECTORY_AUTHORITY` env var is not set.'
```

File: packages/api/src/auth/decoders/azureActiveDirectory.ts
Comment (Tobbe) [inline packages/api/src/auth/decoders/azureActiveDirectory.ts:15]: JSDoc constructs, like `@docs` usually don't work in `//` comments. Change to `/** */`

File: packages/api/src/auth/decoders/azureActiveDirectory.ts
Comment (Tobbe) [inline packages/api/src/auth/decoders/azureActiveDirectory.ts:4]: Do you need to export this?

File: packages/api/src/auth/decoders/azureActiveDirectory.ts
Comment (Tobbe) [inline packages/api/src/auth/decoders/azureActiveDirectory.ts:48]: Any reason you didn't just inline all the code from that function?

File: packages/auth/src/authClients/azureActiveDirectory.ts
Comment (Tobbe) [inline packages/auth/src/authClients/azureActiveDirectory.ts:16]:
```suggestion
    login: async (options?) => await client.loginPopup(options),
```

File: packages/auth/src/authClients/azureActiveDirectory.ts
Comment (Tobbe) [inline packages/auth/src/authClients/azureActiveDirectory.ts:20]:
```suggestion
    signup: async (options?) => await client.loginPopup(options),
```

File: packages/auth/src/authClients/azureActiveDirectory.ts
Comment (Tobbe) [inline packages/auth/src/authClients/azureActiveDirectory.ts:23]:
```suggestion
    getToken: async () => sessionStorage.getItem('msal.idtoken'),
```

File: packages/auth/src/authClients/azureActiveDirectory.ts
Comment (Tobbe) [inline packages/auth/src/authClients/azureActiveDirectory.ts:13]: Do you need/want to specify this one?

File: packages/auth/src/authClients/azureActiveDirectory.ts
Comment (Tobbe) [inline packages/auth/src/authClients/azureActiveDirectory.ts:27]:
```suggestion
    getUserMetadata: async () => (await client.getAccount()) || null,
```

**Review by Tobbe — COMMENTED** [review]
Now that I've worked with this code (https://github.com/redwoodjs/redwood/pull/1536) I feel I can read this and provide some feedback. Mostly nitpicking comments, but since you'll need to handle merging this with the new Supabase stuff anyway, I figured you could take a look at my minor comments as well.

Another thing you could do, that I forgot when I did the Supabase stuff, is update the comment at the end of the auth README, with the list of auth providers (please add Supabase as well if you do 🙂)

If you fix the comments and rebase your code I'll make sure we get this merged. No need to wait for @thedavidprice's `setup` stuff. We can move this over to `setup` later together with every other auth generator.

File: packages/api/src/auth/decoders/azureActiveDirectory.ts
Comment (jeliasson) [reply packages/api/src/auth/decoders/azureActiveDirectory.ts]: Fixed in https://github.com/jeliasson/redwood/commit/0a9d8e922ff37ca10dd620e82cb57606c6f41e56

File: packages/api/src/auth/decoders/azureActiveDirectory.ts
Comment (jeliasson) [reply packages/api/src/auth/decoders/azureActiveDirectory.ts]: Fixed in https://github.com/jeliasson/redwood/commit/6d7044c9d1446aec3ac838e03760f97b35296976

File: packages/auth/src/authClients/azureActiveDirectory.ts
Comment (jeliasson) [reply packages/auth/src/authClients/azureActiveDirectory.ts]: I can't find any related method in [MSAL's UserAgentApplication](https://pub.dev/documentation/msal_js/latest/msal_js/UserAgentApplication-class.html#constructors). Removed in https://github.com/jeliasson/redwood/commit/d458839a1606ed5b174393bc73595ad0a8dac438

File: packages/api/src/auth/decoders/azureActiveDirectory.ts
Comment (jeliasson) [reply packages/api/src/auth/decoders/azureActiveDirectory.ts:48]: I took a shortcut of using [Auth0 as implementation example](https://github.com/redwoodjs/redwood/blob/main/packages/api/src/auth/decoders/auth0.ts#L25-L70). I think we can leave it for now, and revisit that with some refactoring on both auth providers in a separate PR.

File: packages/api/src/auth/decoders/azureActiveDirectory.ts
Comment (jeliasson) [reply packages/api/src/auth/decoders/azureActiveDirectory.ts]: Fixed in https://github.com/jeliasson/redwood/commit/a222300bc5282a2fabef17362ff9eefdd572f1b8

File: packages/auth/src/authClients/azureActiveDirectory.ts
Comment (jeliasson) [reply packages/auth/src/authClients/azureActiveDirectory.ts]: Fixed in https://github.com/jeliasson/redwood/commit/5b2caee1228e8c6579d0f202f691041188fe1720

File: packages/auth/src/authClients/azureActiveDirectory.ts
Comment (jeliasson) [reply packages/auth/src/authClients/azureActiveDirectory.ts]: Fixed in https://github.com/jeliasson/redwood/commit/f14a58c31124662ac9df1e715d78cdeea5d14262

File: packages/auth/src/authClients/azureActiveDirectory.ts
Comment (jeliasson) [reply packages/auth/src/authClients/azureActiveDirectory.ts]: Fixed in https://github.com/jeliasson/redwood/commit/5c651a3bb959d5b329a1b99800f3a6a2aa1d5d5d

File: packages/auth/src/authClients/azureActiveDirectory.ts
Comment (jeliasson) [reply packages/auth/src/authClients/azureActiveDirectory.ts]: Fixed in https://github.com/jeliasson/redwood/commit/a5c608ac87daf42b574db1a1bbbeeec4a6620056

**Review by Tobbe — APPROVED** [review]

**Comment by jeliasson** [comment]
Thanks for the help @Tobbe and @dthyresson! ❤️

I'll bring in the canary release to https://github.com/redwoodjs/playground-auth/pull/6 for final testing, and have another glance on the documentation in https://github.com/redwoodjs/redwoodjs.com/pull/407.

/ping @andrew-hwahin @andrewlamyw

**Comment by andrewlamyw** [comment]
Million thanks for the effort!🙏❤️ @jeliasson @Tobbe @dthyresson 

I think this will definitely drives more RedwoodJS adoption for projects which has the requirement to use AzureAD.

**Pull request merged**
